### PR TITLE
Exclude Spam projects count from total_projects count

### LINKED ISSUE #5927: Fix the count of total projects shown on the homepage
On the hompage (https://readthedocs.org/), we show the total count of the projects.

Screenshot from 2019-07-15 14-23-42

This number also includes projects that are identified as spam.
It would be good if we exclude spam projects from it.

### PR DESCRIPTION
**Pull request by Abhi-khandelwal**
Fixes #5927

### CODE REVIEW

File: readthedocs/core/views/__init__.py [hunk 1]
```diff
@@ -42,7 +42,7 @@ def get_context_data(self, **kwargs):
         """Add latest builds and featured projects."""
         context = super().get_context_data(**kwargs)
         context['featured_list'] = Project.objects.filter(featured=True)
-        context['projects_count'] = Project.objects.count()
+        context['projects_count'] = Project.objects.exclude(slug="spam").count()
```
Comment (dojutsu-user) [inline readthedocs/core/views/__init__.py]: I think we want to have something like this:
https://github.com/readthedocs/readthedocs.org/blob/7be01d844b981635822d4df73f614ad2c3d71b8b/readthedocs/projects/views/public.py#L46

**Comment by saadmk11** [comment]
You may want to add tests for these changes. :)

**Comment by Abhi-khandelwal** [comment]
@saadmk11 Can you help me with these test changes, I mean where should I have to make changes.

**Comment by saadmk11** [comment]
you can add something similar to this for `homepage` with projects in [test_views.py](https://github.com/readthedocs/readthedocs.org/blob/master/readthedocs/rtd_tests/tests/test_views.py)

https://github.com/readthedocs/readthedocs.org/blob/9c5a756e4f72ce021e26ecd8145202aaffe65d78/readthedocs/rtd_tests/tests/test_project_views.py#L415-L419

With `self.assertEqual(<expected_number_of_projects>, response.context['projects_count'])`

**Comment by Abhi-khandelwal** [comment]
@saadmk11 I got this, but still I have one confusion, Do I have to make seaparate class for this  or I have to make a function `test_projects_count()` under `RandomPageTests` class?

**Comment by saadmk11** [comment]
you can make something like `HomepageViewTests`

**Comment by Abhi-khandelwal** [comment]
@dojutsu-user , I would like you to review on above changes :)

File: readthedocs/rtd_tests/tests/test_views.py [hunk 2]
```diff
@@ -277,8 +277,8 @@ def test_build_redirect(self, mock):
             r._headers['location'][1],
             '/projects/pip/builds/%s/' % build.pk,
         )
-
-    def test_build_list_includes_external_versions(self):
+
```
Comment (dojutsu-user) [inline readthedocs/rtd_tests/tests/test_views.py]: Remove this space.

File: readthedocs/rtd_tests/tests/test_views.py [hunk 3]
```diff
@@ -277,8 +277,8 @@ def test_build_redirect(self, mock):
             r._headers['location'][1],
             '/projects/pip/builds/%s/' % build.pk,
         )
-
-    def test_build_list_includes_external_versions(self):
+        
+     def test_build_list_includes_external_versions(self):
```
Comment (dojutsu-user) [inline readthedocs/rtd_tests/tests/test_views.py]: here also.

File: readthedocs/rtd_tests/tests/test_views.py [hunk 4]
```diff
@@ -294,5 +294,15 @@ def test_build_list_includes_external_versions(self):
             reverse('builds_project_list', args=[self.pip.slug]),
         )
         self.assertEqual(response.status_code, 200)
-
+
```
Comment (dojutsu-user) [inline readthedocs/rtd_tests/tests/test_views.py]: here too

File: readthedocs/rtd_tests/tests/test_views.py [hunk 5]
```diff
@@ -294,5 +294,15 @@ def test_build_list_includes_external_versions(self):
...
+        
         self.assertIn(external_version_build, response.context['build_qs'])
+
+
+class HomepageViewTests(TestCase):
+
+    def test_projects_count(self):
+        url = reverse('homepage') 
+        response = self.client.get(url)
+        self.assertEqual(response.status_code, 200)
```
Comment (dojutsu-user) [inline readthedocs/rtd_tests/tests/test_views.py]: The tests can be improved.
You can make some dummy users and projects, then mark some users as `banned` and then check the value of `projects_count` to check if the projects with banned users are excluded.
We are using [Django Dynamic Fixture](https://django-dynamic-fixture.readthedocs.io/en/latest/) to make dummy data.

You can find many examples of that in our tests:
https://github.com/readthedocs/readthedocs.org/blob/356d811b740460afe1620435ece6c9993ba480a0/readthedocs/rtd_tests/tests/test_views.py#L203-L204

Comment (Abhi-khandelwal) [reply readthedocs/rtd_tests/tests/test_views.py]: you mean I have to replace queryset at L-306 to dummy data?

Comment (dojutsu-user) [reply readthedocs/rtd_tests/tests/test_views.py]: Would be better to re-write the tests.

Comment (Abhi-khandelwal) [reply readthedocs/rtd_tests/tests/test_views.py]: @dojutsu-user ,Actually I'm not good in writting test cases, can you please help me with this?
as you said that i have to make dummy users and projects  using DDF ,
`self.user_profile = get(UserProfile, banned=False)
self.user = get(User, profile=user_profile)
self.project = get(Project, users=[user])`

Am i doing right ? :)

Comment (dojutsu-user) [reply readthedocs/rtd_tests/tests/test_views.py]: You can write something like this:

```python3
class HomepageViewTests(TestCase):

    def setUp(self):
        self.homepage_url = reverse('homepage')

    def test_projects_count(self):
        user_1 = get(User)
        profile_1 = get(UserProfile, user=user_1, banned=True)

        user_2 = get(User)
        profile_2 = get(UserProfile, user=user_2, banned=True)

        user_3 = get(User)
        profile_3 = get(UserProfile, user=user_3, banned=False)

        project_a = get(Project, users=[user_1])
        project_b = get(Project, users=[user_2])
        project_c = get(Project, users=[user_3])

        spam_proj_count = Project.objects.filter(users__profile__banned=True).count()

        self.assertEqual(spam_proj_count, 2)
        self.assertEqual(User.objects.all().count(), 3)
        self.assertEqual(User.objects.filter(profile__banned=True).count(), 2)

        expected_count = Project.objects.all().count() - spam_proj_count
        response = self.client.get(self.homepage_url)

        self.assertEqual(response.status_code, 200)
        self.assertEqual(expected_count, response.context['projects_count'])
```

Comment (Abhi-khandelwal) [reply readthedocs/rtd_tests/tests/test_views.py]: Thank you @dojutsu-user

File: readthedocs/rtd_tests/tests/test_views.py [hunk 6]
```diff
@@ -277,7 +278,7 @@ def test_build_redirect(self, mock):
             r._headers['location'][1],
             '/projects/pip/builds/%s/' % build.pk,
         )
-
+
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_views.py]:
```suggestion

```

**Pull request merged**